File: GKPageScrollViewObjc/GKPageScrollViewObjc/Demo/VTMagic/GKVTMagicViewController.m
//
//  GKVTMagicViewController.m
//  GKPageScrollViewObjc
//
//  Created by gaokun on 2019/6/6.
//  Copyright © 2019 gaokun. All rights reserved.
//

#import "GKVTMagicViewController.h"
#import "GKBaseListViewController.h"
#import "GKPageScrollView.h"
#import <VTMagic/VTMagic.h>

@interface GKVTMagicViewController ()<GKPageScrollViewDelegate, VTMagicViewDataSource, VTMagicViewDelegate, GKPageTableViewGestureDelegate>

@property (nonatomic, strong) GKPageScrollView  *pageScrollView;

@property (nonatomic, strong) UIView            *headerView;

@property (nonatomic, strong) UIView            *pageView;

@property (nonatomic, strong) VTMagicController *magicController;

@property (nonatomic, strong) NSArray           *titles;
@property (nonatomic, strong) NSArray           *childVCs;

@end

@implementation GKVTMagicViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.gk_navBarAlpha = 0;
    self.gk_navTitle = @"VTMagic使用";
    self.gk_statusBarStyle = UIStatusBarStyleLightContent;
    self.gk_navTitleColor = [UIColor whiteColor];
    
    [self.view addSubview:self.pageScrollView];
    
    [self.pageScrollView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self.view);
    }];
    
    [self.pageScrollView reloadData];
    
    [self.magicController.magicView reloadData];
}

#pragma mark - GKPageScrollViewDelegate
- (UIView *)headerViewInPageScrollView:(GKPageScrollView *)pageScrollView {
    return self.headerView;
}

- (UIView *)pageViewInPageScrollView:(GKPageScrollView *)pageScrollView {
    return self.pageView;
}

- (NSArray<id<GKPageListViewDelegate>> *)listViewsInPageScrollView:(GKPageScrollView *)pageScrollView {
    return self.childVCs;
}

#pragma mark - GKPageTableViewGestureDelegate
- (BOOL)pageTableView:(GKPageTableView *)tableView gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldRecognizeSimultaneouslyWithGestureRecognizer:(UIGestureRecognizer *)otherGestureRecognizer {
    // 禁止UIScrollView左右滑动时，上下左右都可以滑动
    UIScrollView *scrollView = [self.magicController.magicView valueForKey:@"contentView"];
    
    if (otherGestureRecognizer == scrollView.panGestureRecognizer) {
        return NO;
    }
    return [gestureRecognizer isKindOfClass:[UIPanGestureRecognizer class]] && [otherGestureRecognizer isKindOfClass:[UIPanGestureRecognizer class]];
}

#pragma mark - VTMagicViewDataSource
- (NSArray<NSString *> *)menuTitlesForMagicView:(VTMagicView *)magicView {
    return self.titles;
}

- (UIButton *)magicView:(VTMagicView *)magicView menuItemAtIndex:(NSUInteger)itemIndex {
    static NSString *itemIdentifier = @"itemIdentifier";
    UIButton *menuItem = [magicView dequeueReusableItemWithIdentifier:itemIdentifier];
    if (!menuItem) {
        menuItem = [UIButton buttonWithType:UIButtonTypeCustom];
        [menuItem setTitleColor:[UIColor grayColor] forState:UIControlStateNormal];
        [menuItem setTitleColor:[UIColor blackColor] forState:UIControlStateSelected];
        menuItem.titleLabel.font = [UIFont fontWithName:@"Helvetica" size:15.f];
    }
    return menuItem;
}

- (UIViewController *)magicView:(VTMagicView *)magicView viewControllerAtPage:(NSUInteger)pageIndex {
    return self.childVCs[pageIndex];
}

#pragma mark - VTMagicViewDelegate

#pragma mark - 懒加载
- (GKPageScrollView *)pageScrollView {
    if (!_pageScrollView) {
        _pageScrollView = [[GKPageScrollView alloc] initWithDelegate:self];
        _pageScrollView.mainTableView.gestureDelegate = self;
    }
    return _pageScrollView;
}

- (UIView *)headerView {
    if (!_headerView) {
        _headerView = [[UIView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, 200.0f)];
        _headerView.backgroundColor = [UIColor colorWithPatternImage:[UIImage imageNamed:@"test"]];
    }
    return _headerView;
}

- (UIView *)pageView {
    if (!_pageView) {
        _pageView = self.magicController.view;
    }
    return _pageView;
}

- (VTMagicController *)magicController {
    if (!_magicController) {
        _magicController = [[VTMagicController alloc] init];
        _magicController.magicView.navigationColor = [UIColor whiteColor];
        _magicController.magicView.sliderColor = [UIColor redColor];
        _magicController.magicView.layoutStyle = VTLayoutStyleDivide;
        _magicController.magicView.switchStyle = VTSwitchStyleDefault;
        _magicController.magicView.navigationHeight = 40.f;
        _magicController.magicView.dataSource = self;
        _magicController.magicView.delegate = self;
    }
    return _magicController;
}

- (NSArray *)titles {
    if (!_titles) {
        _titles = @[@"详情", @"热门", @"相关", @"聊天"];
    }
    return _titles;
}

- (NSArray *)childVCs {
    if (!_childVCs) {
        GKBaseListViewController *detailVC = [GKBaseListViewController new];
        detailVC.shouldLoadData = NO;
        
        GKBaseListViewController *hotVC = [GKBaseListViewController new];
        hotVC.shouldLoadData = NO;
        
        GKBaseListViewController *aboutVC = [GKBaseListViewController new];
        aboutVC.shouldLoadData = NO;
        
        GKBaseListViewController *chatVC = [GKBaseListViewController new];
        chatVC.shouldLoadData = NO;
        
        _childVCs = @[detailVC, hotVC, aboutVC, chatVC];
    }
    return _childVCs;
}

@end

File: GKPageScrollViewSwift/GKPageScrollViewSwift/Demo/VTMagic/GKVTMagicViewController.swift
//
//  GKVTMagicViewController.swift
//  GKPageScrollViewSwift
//
//  Created by gaokun on 2019/10/18.
//  Copyright © 2019 gaokun. All rights reserved.
//

import UIKit
import VTMagic

class GKVTMagicController: VTMagicController {
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.magicView.navigationColor = UIColor.white
        self.magicView.sliderColor = UIColor.red
        self.magicView.layoutStyle = .divide
        self.magicView.switchStyle = .default
        self.magicView.navigationHeight = 40.0
    }
}

class GKVTMagicViewController: GKDemoBaseViewController {

    lazy var pageScrollView: GKPageScrollView! = {
        let pageScrollView = GKPageScrollView(delegate: self)
        pageScrollView.mainTableView.gestureDelegate = self
        return pageScrollView
    }()
    
    lazy var headerView: UIView! = {
        let headerView = UIView(frame: CGRect(x: 0, y: 0, width: view.frame.size.width, height: 200.0))
        headerView.backgroundColor = UIColor(patternImage: UIImage(named: "test")!)
        return headerView
    }()
    
    lazy var magicVC: VTMagicController = {
        let magicVC = VTMagicController()
        magicVC.magicView.navigationColor = UIColor.white
        magicVC.magicView.sliderColor = UIColor.red
        magicVC.magicView.layoutStyle = .divide
        magicVC.magicView.switchStyle = .default
        magicVC.magicView.navigationHeight = 40.0
        magicVC.magicView.dataSource = self
        magicVC.magicView.delegate = self
        return magicVC
    }()
    
    lazy var pageView: UIView! = {
        let pageView = self.magicVC.view
        return pageView
    }()
    
    let titles = ["详情", "热门", "相关", "聊天"]
    lazy var childVCs: [GKBaseListViewController] = {
        var childVCs = [GKBaseListViewController]()
        
        let detailVC = GKBaseListViewController()
        detailVC.shouldLoadData = false
        childVCs.append(detailVC)
        
        let hotVC = GKBaseListViewController()
        hotVC.shouldLoadData = false
        childVCs.append(hotVC)
        
        let aboutVC = GKBaseListViewController()
        aboutVC.shouldLoadData = false
        childVCs.append(aboutVC)
        
        let chatVC = GKBaseListViewController()
        chatVC.shouldLoadData = false
        childVCs.append(chatVC)
        
        return childVCs
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.gk_navBarAlpha = 0
        self.gk_navTitle = "VTMagic使用"
        self.gk_statusBarStyle = .lightContent
        self.gk_navTitleColor = UIColor.white
        
        self.view.addSubview(self.pageScrollView)
        
        self.pageScrollView.snp.makeConstraints { (make) in
            make.edges.equalTo(self.view)
        }
        
        self.pageScrollView.reloadData()
        self.magicVC.magicView.reloadData()
    }
}

extension GKVTMagicViewController: GKPageScrollViewDelegate {
    func headerView(in pageScrollView: GKPageScrollView) -> UIView {
        return self.headerView
    }
    
    func pageView(in pageScrollView: GKPageScrollView) -> UIView {
        return self.pageView
    }
    
    func listView(in pageScrollView: GKPageScrollView) -> [GKPageListViewDelegate] {
        return self.childVCs
    }
}

extension GKVTMagicViewController: VTMagicViewDataSource, VTMagicViewDelegate {
    func menuTitles(for magicView: VTMagicView) -> [String] {
        return self.titles
    }
    
    func magicView(_ magicView: VTMagicView, menuItemAt itemIndex: UInt) -> UIButton {
        let identifier = "itemIdentifier"
        var menuItem = magicView.dequeueReusableItem(withIdentifier: identifier)
        if menuItem == nil {
            menuItem = UIButton(type: .custom)
            menuItem?.setTitleColor(UIColor.gray, for: .normal)
            menuItem?.setTitleColor(UIColor.black, for: .selected)
            menuItem?.titleLabel?.font = UIFont(name: "Helvetica", size: 15.0)
        }
        return menuItem!
    }
    
    func magicView(_ magicView: VTMagicView, viewControllerAtPage pageIndex: UInt) -> UIViewController {
        return self.childVCs[Int(pageIndex)]
    }
}

extension GKVTMagicViewController: GKPageTableViewGestureDelegate {
    func pageTableView(_ tableView: GKPageTableView, gestureRecognizer: UIGestureRecognizer, shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer) -> Bool {
        
        let scrollView: UIScrollView = self.magicVC.magicView.value(forKey: "contentView") as! UIScrollView
        
        if otherGestureRecognizer == scrollView.panGestureRecognizer {
            return false
        }
        
        return gestureRecognizer.view?.isKind(of: UIScrollView.classForCoder()) ?? false && otherGestureRecognizer.view?.isKind(of: UIScrollView.classForCoder()) ?? false
    }
}
